engine/api/workflow/repository.go

```diff
@@ -0,0 +1,226 @@
+	// Read files
+	tr, err := ReadCDSFiles(ope.LoadFiles.Results)
+	if err != nil {
+		return nil, sdk.WrapError(err, "extractWorkflow> Unable to read cds files")
+	}
+	opt := &PushOption{
+		VCSServer:          ope.VCSServer,
+		RepositoryName:     ope.RepositoryInfo.Name,
+		RepositoryStrategy: ope.RepositoryStrategy,
+		Branch:             ope.Setup.Checkout.Branch,
+		FromRepository:     ope.RepositoryInfo.FetchURL,
+		IsDefaultBranch:    ope.Setup.Checkout.Branch == ope.RepositoryInfo.DefaultBranch,
+		DryRun:             true,
+		HookUUID:           hookUUID,
+	}
+
+	allMsg, workflowPushed, errP := Push(db, store, p, tr, opt, u, decryptFunc)
+	if errP != nil {
+		return nil, sdk.WrapError(errP, "extractWorkflow> Unable to get workflow from file")
+	}
+	*w = *workflowPushed
+	return allMsg, nil
+}
+
+// ReadCDSFiles reads CDS files
+func ReadCDSFiles(files map[string][]byte) (*tar.Reader, error) {
+	// Create a buffer to write our archive to.
+	buf := new(bytes.Buffer)
+	// Create a new tar archive.
+	tw := tar.NewWriter(buf)
+	// Add some files to the archive.
+	for fname, fcontent := range files {
+		log.Debug("ReadCDSFiles> Reading %s", fname)
+		hdr := &tar.Header{
+			Name: filepath.Base(fname),
+			Mode: 0600,
+			Size: int64(len(fcontent)),
+		}
+		if err := tw.WriteHeader(hdr); err != nil {
+			return nil, sdk.WrapError(err, "ReadCDSFiles> Cannot write header")
+		}
+		if n, err := tw.Write(fcontent); err != nil {
+			return nil, sdk.WrapError(err, "ReadCDSFiles> Cannot write content")
+		} else if n == 0 {
+			return nil, fmt.Errorf("nothing to write")
+		}
+	}
+	// Make sure to check the error on Close.
+	if err := tw.Close(); err != nil {
+		return nil, err
+	}
+
+	return tar.NewReader(buf), nil
+}
+
+func pollRepositoryOperation(c context.Context, db gorp.SqlExecutor, store cache.Store, ope *sdk.Operation) error {
+	tickTimeout := time.NewTicker(10 * time.Minute)
+	tickPoll := time.NewTicker(2 * time.Second)
+	defer tickTimeout.Stop()
+	for {
+		select {
+		case <-c.Done():
+			if c.Err() != nil {
+				return sdk.WrapError(c.Err(), "pollRepositoryOperation> Exiting")
+			}
+		case <-tickTimeout.C:
+			return sdk.WrapError(sdk.ErrRepoOperationTimeout, "pollRepositoryOperation> Timeout analyzing repository")
+		case <-tickPoll.C:
+			if err := GetRepositoryOperation(db, store, ope); err != nil {
+				return sdk.WrapError(err, "pollRepositoryOperation> Cannot get repository operation status")
+			}
+			switch ope.Status {
+			case sdk.OperationStatusError:
+				return sdk.WrapError(fmt.Errorf("%s", ope.Error), "getImportAsCodeHandler> Operation in error. %+v", ope)
+			case sdk.OperationStatusDone:
+				return nil
+			}
+			continue
+		}
+	}
+}
+
+func createOperationRequest(w sdk.Workflow, opts sdk.WorkflowRunPostHandlerOption) (sdk.Operation, error) {
+	ope := sdk.Operation{}
+	if w.Root.Context.Application == nil {
```

Comment:
just to check, but here the context could be nil

Reply:
Context cannot be nil after loading workflow from DB, each node must have a context